soc: arm: st_stm32: stm32l4: power.c standby shutdown mode

Desvauxm-st opened the pull request:
STM32L4x power management of Standby mode
and shutdown mode

Desvauxm-st committed:
soc: arm: st_stm32: stm32l4: power.c standby shutdownn mode

STM32L4x power management stop mode modification


Signed-off-by: Marc Desvaux <marc.desvaux-ext@st.com>

FRASTM reviewed (changes requested):
* go with 2 commits, at least ; one for sample one  for soc 
* fix CI warnings
* detail the  correlation with sleep "K_TICKS_FOREVER" if any

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -20,89 +20,139 @@
+void set_mode_standby(void)
+{
+	/* Select standby mode */
+	printk(" mode_standby done\n\n\n");
```
No printk. Replace by LOG_BDG

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,154 @@
+	if ((LL_PWR_IsActiveFlag_SB() == true) && (cause == 0))	{
+		LL_PWR_ClearFlag_SB();
+		LL_PWR_ClearFlag_WU();
+		printk(" reset from Standby mode exit\n");
+		printk(" reset from Standby mode exit\n");
```
```suggestion
		printk("reset from Standby mode exit\n");
```
Apply everywhere

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,154 @@
+
+	if (cause == (RESET_PIN | RESET_BROWNOUT)) {
+		LL_PWR_ClearFlag_WU();
+		printk(" reset from shutdown mode exit or power up\n");
+		printk(" reset from shutdown mode exit or power up\n");
```
Why duplicating all printk ?

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,154 @@
+
+
+#define SLEEP_TIME_MS   300
+
+
```
#if !defined(CONFIG_SOC_SERIES_STM32L4X)
#error Not implemented for other series
#endif

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
```
Fix copyright

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,147 @@
+#error "Unsupported board: sw0 devicetree alias is not defined"
+#endif
+
+#if !defined(CONFIG_SOC_SERIES_STM32L4X)
+#error Not implemented for other series
+#endif  /* CONFIG_SOC_SERIES_STM32L4X */
```
Move this earlier

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,147 @@
+#error Not implemented for other series
+#endif  /* CONFIG_SOC_SERIES_STM32L4X */
+
+static const struct gpio_dt_spec button = GPIO_DT_SPEC_GET_OR(SW0_NODE, gpios,
+							      {0});
+
+
+static const struct gpio_dt_spec led =
+	GPIO_DT_SPEC_GET(DT_ALIAS(led0), gpios);
```
Use consistent indentation

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,147 @@
+	/* Clear wakeup flags */
+	LL_PWR_ClearFlag_WU();
+}
+
+
+
+
```
Remove superflous lines.
Apply everywhere

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,147 @@
+
+void main(void)
+{
+
+
```
Superfluous lines

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,147 @@
+
+	led_is_on = true;
+	while (true) {
+		int i;
+
+		for (i = 1 ; i < 10 ; i++) {
```
Is the for loop really required ?

erwango replied:
Please use GPIO interrupt if reactivity required

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -20,89 +20,140 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
 
+
+
```
Remove these lines

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -20,89 +20,140 @@
 
-/* Invoke Low Power/System Off specific Tasks */
-__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+
+
```
remove these lines

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -20,89 +20,140 @@
-	if (state != PM_STATE_SUSPEND_TO_IDLE) {
-		LOG_DBG("Unsupported power state %u", state);
-		return;
+
```
remove this line

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -20,89 +20,140 @@
+	/* Configure CPU core */
+	/* Enable CPU deep sleep mode */
+#ifndef CONFIG_DEBUG
+	LL_LPM_EnableDeepSleep();
```
Is this required before shutdown/standby ?

Desvauxm-st replied:
yes

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -20,89 +20,140 @@
+	default:
+		LOG_DBG("Unsupported power state substate-id %u", substate_id);
+		break;
+	}
```
k_cpu_idle();

Desvauxm-st replied:
k_cpu_idle(); move in pm_state_set() only for PM_STATE_SUSPEND_TO_IDLE

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,167 @@
+	int sleep_ms;
+	sleep_ms = 300;
+	k_sem_init(&button_sem, 0, 1);
+	while (true) {
```
I don't think while loop is required

Desvauxm-st replied:
OK

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,167 @@
+				k_msleep(200);
+			}
+		}
+		if (led_is_on == true) {
```
else

Desvauxm-st replied:
ok

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,167 @@
+	while (true) {
+		k_sem_take(&button_sem, K_FOREVER);
+		if (led_is_on == false) {
+			gpio_pin_configure(led.port, led.pin, GPIO_DISCONNECTED);
+			printk("user_button pressed\n");
+			printk("enter into Shutdown Mode\n");
+			printk("release the user button to exit from Shutdown Mode\n\n");
```
Let's factorize when possible

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,168 @@
+			config_wakeup_features();
+			pm_state_force(0u, &(struct pm_state_info) {PM_STATE_STANDBY, 0, 0});
+
+			k_msleep(200);
+			while (1) {
+			/* stay in Standby mode until wakeup line activated */
+				k_msleep(200);
+			}
```
This shouldn't be required. Can you have a new try w/o ?

Desvauxm-st replied:
Yes right

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,168 @@
+			printk("user_button pressed\n");
+			printk("enter into Shutdown Mode\n");
+			printk("release the user button to exit from Shutdown Mode\n\n");
+			config_wakeup_features();
```
This doesn't have to be done after button press.
Can you move it earlier ?

Desvauxm-st replied:
yes

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,168 @@
+	hwinfo_get_reset_cause(&cause);
+	hwinfo_clear_reset_cause();
+
+	if ((LL_PWR_IsActiveFlag_SB() == true) && (cause == 0))	{
```
It would be great to have LL_PWR_IsActiveFlag_SB detection as part of `hwinfo_get_reset_cause`.
Can you check if possible to do this ?
(This would be another PR)

Desvauxm-st replied:
Yes I will try to check if it possible to do that

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,168 @@
+	hwinfo_clear_reset_cause();
+
+	if ((LL_PWR_IsActiveFlag_SB() == true) && (cause == 0))	{
+		LL_PWR_ClearFlag_SB();
+		LL_PWR_ClearFlag_WU();
```
Should it be done in `hwinfo_clear_reset_cause();` ?
(to be done in the hwinfo PR)

Desvauxm-st replied:
Yes I will try to check if it possible to do that

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,168 @@
+	gpio_init_callback(&button_cb_data, button_pressed, BIT(button.pin));
+	gpio_add_callback(button.port, &button_cb_data);
+
+	printk(" Device ready :%s\n\n\n", CONFIG_BOARD);
```
```suggestion
	printk("Device ready: %s\n\n\n", CONFIG_BOARD);
```

Desvauxm-st replied:
ok

erwango reviewed (changes requested):
Some more changes requested, but globally looks good.

Regarding the app, can you work on readability?
For instance, following would be easier to read:
```
*** Booting Zephyr OS build zephyr-v3.3.0-619-g693bf8e46f4b ***   
                                                                  
Reset cause: Shutdown Mode                                            
                                                                  
Device ready: b_l4s5i_iot01a                                      
                                                                  
Press and hold the user button:                                   
  when LED2 is OFF to enter to Shutdown Mode                      
  when LED2 is ON to enter to Standby Mode                        
                                                                  
User button pressed: 
Standby Mode requested         
                                 
Release the user button to exit from Standby Mode                 
                                                                  
*** Booting Zephyr OS build zephyr-v3.3.0-619-g693bf8e46f4b ***   

```

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+    harness: console
+    harness_config:
+      type: one_line
+      regex:
+        - "Device ready"
```
This will fail in CI. 
First message when running twister is likely `Reset cause: Reset pin`

Desvauxm-st replied:
ok

erwango reviewed (changes requested):
Few more comments.

Last: Please split in 3 commits:
- Update power.c to factorise stop states
- Update power.c to add standby and shutdown
- Add the sample

Also rename the sample to: standby_shutdown

Thanks

erwango commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,153 @@
+		pm_state_force(0u, &(struct pm_state_info) {PM_STATE_STANDBY, 0, 0});
+		/* stay in Standby mode until wakeup line activated */
+	}
+	k_msleep(200);
```
Is that strictly required ?

Desvauxm-st replied:
no works without k_msleep

Desvauxm-st commented on samples/boards/stm32/power_mgmt/blinky_standby_shutdown_mode/src/main.c:
```diff
@@ -0,0 +1,153 @@
+#define PRIORITY 7
+#define SLEEP_TIME_MS   3000
+
+#include <stm32_ll_pwr.h>
```
yes

Desvauxm-st committed:
soc: arm: st_stm32: stm32l4: power.c ultra_low_power mode

STM32L4x power management (ultra_low_power) of Standby mode
and shutdown mode ultra_low_power

Signed-off-by: Marc Desvaux <marc.desvaux-ext@st.com>

Desvauxm-st committed:
samples: boards: stm32: Power_mgmt: sample for standby shutdownn mode

STM32L4x power management (ultra_low_power) of Standby mode
and shutdown mode

Signed-off-by: Marc Desvaux <marc.desvaux-ext@st.com>

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -27,82 +27,93 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+	switch (state) {
+	case PM_STATE_SUSPEND_TO_IDLE:
+		set_mode_stop(substate_id);
+	/* Set SLEEPDEEP bit of Cortex System Control Register */
```
Fix comment indentation

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -54,6 +70,22 @@ void set_mode_stop(uint8_t substate_id)
+void set_mode_standby(void)
+{
+	/* Select standby mode */
+	LOG_DBG("mode_standby done\n\n\n");
```
Otherwise, remove these logs.

erwango commented on samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -0,0 +1,152 @@
+
+void thread_Shutdown_Standby_mode(void)
+{
+	int sleep_ms;
+
+	sleep_ms = 300;
```
Not required anymore

Desvauxm-st replied:
ok

Desvauxm-st commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -27,82 +27,93 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+	/* Set SLEEPDEEP bit of Cortex System Control Register */
 		LL_LPM_EnableDeepSleep();
-		/* enter SLEEP mode : WFE or WFI */
+	/* Select mode entry : WFE or WFI and enter the CPU selected mode */
```
ok

Desvauxm-st commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -27,6 +27,22 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
 #define RCC_STOP_WAKEUPCLOCK_SELECTED LL_RCC_STOP_WAKEUPCLOCK_HSI
 #endif
 
+void enter_low_power_mode(void)
```
ok

Desvauxm-st commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -27,6 +27,22 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+	/* Configure CPU core */
+	/* Enable CPU deep sleep mode */
+#ifndef CONFIG_DEBUG
+	LL_LPM_EnableDeepSleep();
```
ok

erwango commented on samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -0,0 +1,155 @@
+	if (led_is_on == false) {
+		printk("Shutdown Mode requested\n");
+		printk("Release the user button to exit from Shutdown Mode\n\n");
+#ifdef CONFIG_LOG
+		k_msleep(2000);
+#endif /* CONFIG_LOG */
```
2 sec seems quite long for user. Likely user would release the button before that.
Also 2 sec seems large to print a 30 char line, what about 200 ms  ?

erwango commented on samples/boards/stm32/power_mgmt/standby_shutdown/sample.yaml:
```diff
@@ -0,0 +1,19 @@
+      type: multi_line
+      regex:
+        - "Reset cause: Reset pin"
+        - "Device ready: nucleo_l476rg"
```
- "Device ready: .*"

erwango commented on samples/boards/stm32/power_mgmt/standby_shutdown/sample.yaml:
```diff
@@ -0,0 +1,19 @@
+  name: STM32 GPIO Power Management
+tests:
+  sample.boards.stm32.power_mgmt.standby_shutdown:
+    platform_allow: nucleo_l476rg
```
It would be nice to get this working on disco_l475_iot1 which is the L4 target used on bench.
Can we instead filter on CONFIG_SOC_SERIES_STM32L4X ?
(I guess the only condition is to know if config_wakeup_features is also valid for most other L4 targets)

erwango commented on samples/boards/stm32/power_mgmt/standby_shutdown/README.rst:
```diff
@@ -0,0 +1,49 @@
+*****************
+
+By default, :kconfig:option:`CONFIG_PM_DEVICE` and `CONFIG_HWINFO` are enabled.
+for another board than nucleo_L476RG please adjust wakeup pin into config_wakeup_features()
```
Move this in Requirements

erwango commented on samples/boards/stm32/power_mgmt/standby_shutdown/README.rst:
```diff
@@ -0,0 +1,49 @@
+when LED2 is ON to enter to Standby Mode
+release the user button to exit from shutdown mode or from shutdown mode.
+
+
```
Remove this line

FRASTM commented on samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -0,0 +1,155 @@
+		printk("Shutdown Mode requested\n");
+		printk("Release the user button to exit from Shutdown Mode\n\n");
+#ifdef CONFIG_LOG
+		k_msleep(2000);
```
why sleeping for 2s when CONFIG_LOG ?

FRASTM reviewed (commented):
LGTM

tagunil commented on samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:
```diff
@@ -0,0 +1,155 @@
+	k_sem_give(&button_sem);
+}
+
+void thread_Shutdown_Standby_mode(void)
```
Shouldn't that be lowercase?

tagunil commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -27,82 +27,124 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+	LL_DBGMCU_DisableDBGStandbyMode();
+	/* Enter ultra_low-power mode */
+
+	for (;;) {
```
Is that actually better than just calling k_cpu_idle()?

tagunil replied:
Sorry, I haven't been clear enough. The question was not about the loop, but about adding a separate wfi path instead of going through k_cpu_idle() which also does wfi.

Desvauxm-st replied:
OK I'll take a look and test if this change works.

tagunil replied:
I mean, why not just adapt the approach used in stm32u5/power.c? I think the code could be almost identical.

tagunil commented:
@erwango @Desvauxm-st @FRASTM what do you think about using the STM32U5 power management implementation as a base for the shared one? I think the implementation could be unified at least for L0/L1/L4/L5/G0/G4/U5 using some ifdefs. Not sure about dual-core ones.

erwango commented:
> @erwango @Desvauxm-st @FRASTM what do you think about using the STM32U5 power management implementation as a base for the shared one? I think the implementation could be unified at least for L0/L1/L4/L5/G0/G4/U5 using some ifdefs. Not sure about the dual-core ones.

I'm not strongly opposed to it. This would bring some benefits indeed. My main request would be to keep it the less cluttered possible and easy to opt-out (I have in mind some upcoming series which would require real divergence from the main scheme). Are you planning to work on that ?

tagunil commented:
> > @erwango @Desvauxm-st @FRASTM what do you think about using the STM32U5 power management implementation as a base for the shared one? I think the implementation could be unified at least for L0/L1/L4/L5/G0/G4/U5 using some ifdefs. Not sure about the dual-core ones.
> 
> I'm not strongly opposed to it. This would bring some benefits indeed. My main request would be to keep it the less cluttered possible and easy to opt-out (I have in mind some upcoming series which would require real divergence from the main scheme). Are you planning to work on that ?

I do. One of my new projects requires consistent power management support across several series, so I think common implementation could be useful for that.
